shelley/chain-and-ledger/executable-spec/src/PoolRank.hs

```diff
@@ -0,0 +1,93 @@
+module PoolRank
+  ( desirability
+  , rankedPools
+  ) where
+
+import           BaseTypes (intervalValue)
+import           Coin (Coin (..))
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Ratio ((%))
+import qualified Data.Set as Set
+import           EpochBoundary (BlocksMade (..), SnapShots (..), Stake (..), poolStake)
+import           Keys (KeyHash)
+import           LedgerState (AccountState (..), EpochState (..), NewEpochState (..))
+import           Numeric.Natural (Natural)
+import           PParams (PParams (..))
+import           Slot (EpochSize)
+import           TxData (PoolParams (..))
+
+
+-- | Desirability calculation for non-myopic utily,
+-- corresponding to f^~ in section 5.6.1 of
+-- "Design Specification for Delegation and Incentives in Cardano"
+desirability
+  :: PParams
+  -> Coin
+  -> Natural
+  -> Natural
+  -> PoolParams crypto
+  -> Stake crypto
+  -> Coin
+  -> Rational
+desirability pp r blocksN blocksTotal pool (Stake stake) (Coin total) =
+  if fTilde <= cost
+    then 0
+    else (fTilde - cost)*(1 -margin)
+  where
+    fTilde      = fTildeNumer / fTildeDenom
+    fTildeNumer = pBar * fromIntegral r * (z0 + min s z0 * a0)
+    fTildeDenom = 1 + a0
+
+    cost   = fromIntegral $ _poolCost pool
+    margin = intervalValue $ _poolMargin pool
+    tot = max 1 (fromIntegral total)
+
+    Coin pledge = _poolPledge pool
+    s = fromIntegral pledge % tot
+    a0 = _a0 pp
+    z0 = 1 % max 1 (fromIntegral (_nOpt pp))
+
+    pBar = if intervalValue (_d pp) < 0.8 && sigma > 0
+             then beta / sigma
+             else 1
+    Coin pstake = Map.foldl (+) (Coin 0) stake
+    sigma = fromIntegral pstake % tot
+    beta = fromIntegral blocksN / fromIntegral (max 1 blocksTotal)
+
+-- | Computes the top ranked stake pools
+-- corresponding to section 5.6.1 of
+-- "Design Specification for Delegation and Incentives in Cardano"
+rankedPools
+  :: NewEpochState crypto
+  -> EpochSize
+  -> Map (KeyHash crypto) Rational
+rankedPools nes slotsPerEpoch = rankings
+  where
+    EpochState acnt ss _ pp = nesEs nes
+    poolParams = _poolsSS ss
+    (stake@(Stake stake'), delegs) = _pstakeGo ss
+    (BlocksMade b) = nesBprev nes
+
+    -- compute r
+    Coin reserves = _reserves acnt
+    blocksMade = fromIntegral $ sum b :: Integer
+    expectedBlocks = intervalValue (_activeSlotCoeff pp) * fromIntegral slotsPerEpoch
+    eta = fromIntegral blocksMade / min 1 expectedBlocks
+    r = floor $ min 1 eta * intervalValue (_rho pp) * fromIntegral reserves
```

Comment:
That's only the rewards from monetary expansion, right? Should we include also the contributions from fees and decaying deposits? I do not think that's necessary, since we are interested in the ordering of the pools, and this would just be a global factor. Is that right?

Reply:
oh right, I did miss the fees. I was actually going to bring up that very point about `R` (or `r` here, since variables start lowercase :) ).

The _only_ thing that calculating `R` does is allow us to give every pool that cannot cover their cost a desirability score of zero. It is exactly as you say, a constant factor. Setting these pools to zero just means that this score is always non-negative and that we cannot distinguish between pools that cannot cover their costs. The number is probably more meaningful (ie translatable into expectations of rewards), though, if you include R. But it does make the calculation have to touch a lot more state.

Reply:
the calculation of `R` above also does not remove the portion that goes to the treasury. let's figure out to how helpful this constant is in the desirability score, or if the ranking is all that really matters. then I can either remove R altogether or make it more accurate.

Reply:
Oh right. So it _could_ make a difference, if the rewards from monetary expansion alone do not cover the costs of a pool, but the overall rewards, including fees, would. My expectation is that that shouldn't matter too much, since the costs of running a pool will be pretty low anyway, so I would tend to keep it simple here, considering that it's not affecting the ledger state itself.